New input jboss plugin

toni-moreno opened the pull request:
Resolves #824 

### Required for all PRs:

- [X] Signed [CLA](https://influxdata.com/community/cla/).
- [X] Associated README.md updated.
- [X] Has appropriate unit tests.

### New JBoss Plugin 

Based on the great work from @stefa975 in (https://github.com/stefa975/telegraf/tree/master/plugins/inputs/jboss )  fix  (https://github.com/influxdata/telegraf/issues/824)  with some fixes and improvements.

Tested on standalone and Domain Based JBoss 6.X and 7.X on working ok on producction servers.

toni-moreno commented:
Hi @danielnelson When checking the failed test I've seen another input  plugin (go-hddtemp) failing. Can you help me to review please?
````
ok  	github.com/influxdata/telegraf/plugins/inputs/hddtemp/go-hddtemp	1.016s
2017/12/03 20:43:56 I! Started HTTP listener service on :0
2017/12/03 20:43:56 http: TLS handshake error from 127.0.0.1:54011: remote error: tls: bad certificate
2017/12/03 20:43:56 I! Stopped HTTP listener service on  :0
--- FAIL: TestWriteHTTPSNoClientAuth (0.02s)
	<autogenerated>:1: 
                          
	Error Trace:	http_listener_test.go:233
		
	Error:      	Received unexpected error:
		
	            	Post https://localhost:37767/write?db=mydb: x509: certificate has expired or is not yet valid
2017/12/03 20:43:56 I! Started HTTP listener service on :0
2017/12/03 20:43:56 http: TLS handshake error from 127.0.0.1:37016: remote error: tls: bad certificate
2017/12/03 20:43:56 I! Stopped HTTP listener service on  :0
--- FAIL: TestWriteHTTPSWithClientAuth (0.01s)
	<autogenerated>:1: 
                          
	Error Trace:	http_listener_test.go:247
		
	Error:      	Received unexpected error:
		
	            	Post https://localhost:39733/write?db=mydb: x509: certificate has expired or is not yet valid
````

danielnelson commented:
I think the test failure will be resolved if you rebase.

toni-moreno commented:
Hi @danielnelson after rebase the test has been passed ok

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+var sampleConfig = `
+  # Config for get statistics from JBoss AS
+  servers = [
+    "http://[jboss-server-ip]:9090/management",
```
Use `localhost` as the default here.

toni-moreno replied:
OK! I will fix ASAP

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+  servers = [
+    "http://[jboss-server-ip]:9090/management",
+  ]
+	## Execution Mode
```
Adjust indention to be 2 spaces

toni-moreno replied:
OK done!

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+}
+
+func (h *JBoss) checkAuth(host string, uri string) error {
+	url := h.Servers[0]
```
Always uses the first server?

toni-moreno replied:
This method belongs to the legacy code from previous contributor, and I'm right now just reviewing it. seems no longer used ... still reviewing

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		panic(err)
```
Don't panic, chain errors up and out of Gather.

toni-moreno replied:
just the same not used method

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+	return "Telegraf plugin for gathering metrics from JBoss AS"
+}
+
+func (h *JBoss) checkAuth(host string, uri string) error {
```
Where is this function called from?

toni-moreno replied:
yes !!! not  used !!!

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+	return fmt.Sprintf("%x", b)[:16]
+}
+
+func getDigestAuthrization(digestParts map[string]string) string {
```
Typo in function name: Authrization -> Authorization.

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+	return authorization
+}
+
+func getDigestAuthrization2(digestParts map[string]string) string {
```
Should be named more descriptively, also looks like this is the only one used, so maybe remove the other implementation?

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+			return err
+		}
+		tr := &http.Transport{
+			ResponseHeaderTimeout: time.Duration(3 * time.Second),
```
Remove this timeout, client timeout is enough (sorry there are lots of examples doing this I know...)

toni-moreno replied:
ok

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+		if err != nil {
+			return err
+		}
+		tr := &http.Transport{
```
Can you add `Proxy: http.ProxyFromEnvironment,`, I will probably be adding this to many plugins soon.

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+		h.client.SetHTTPClient(client)
+	}
+
+	errorChannel := make(chan error, len(h.Servers))
```
Use acc.AddError instead of an error channel.

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+) error {
+	var wg sync.WaitGroup
+
+	errorChannel := make(chan error, len(hosts))
```
Use `acc.AddError`

danielnelson commented on plugins/inputs/jboss/README.md:
```diff
@@ -0,0 +1,149 @@
+Please refer to JBoss management API for full documentation, https://docs.jboss.org/author/display/AS71/The+HTTP+management+API
+
+
+### Measurements & Fields:
```
Please check the example readme and update the docs to match: https://github.com/influxdata/telegraf/blob/master/plugins/inputs/EXAMPLE_README.md

toni-moreno replied:
Sorry I didn't see this comment !!! , now done

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+		return nil, err
+	}
+
+	//req, err := http.NewRequest("POST", serverUrl.String(), bytes.NewBuffer(requestBody))
```
Clean up unused code.

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+	method := "POST"
+
+	// Debug JSON request
+	log.Printf("D! Req: %s\n", requestBody)
```
I feel like the debug logging in this function to too sloppy, we can leave it in if we make sure it is readable and understandable, otherwise remove it.

toni-moreno replied:
Yes it is , and helped me to understand all the jboss management protocol , iture  will help if somebody has a particular jboss installation and having troubles gathering certain metrics.

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+	}
+}
+
+//func (j *JBoss) prepareRequest(optype int, adress map[string]interface{}) (map[string]interface{}, error) {
```
Remove this line

toni-moreno replied:
ok

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+}
+
+//func (j *JBoss) prepareRequest(optype int, adress map[string]interface{}) (map[string]interface{}, error) {
+func (h *JBoss) prepareRequest(optype int, adress OrderedMap) (map[string]interface{}, error) {
```
Function name seem wrong, should be something more like: `createRequestBody`

typo adress -> address

toni-moreno replied:
ok!

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+		bodyContent["operation"] = "read-children-names"
+		bodyContent["child-type"] = "host"
+		bodyContent["address"] = []string{}
+		bodyContent["json.pretty"] = 1
```
Probably don't want pretty printed json.

toni-moreno replied:
yes, I will rewrite thisi bit of code

danielnelson commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1222 @@
+	bodyContent := make(map[string]interface{})
+
+	// Create bodyContent
+	switch optype {
```
This seems very verbose, I think there is potential to reduce the amount of code in this switch block.

toni-moreno replied:
As I said before , the jboss management API is complex enough, and each request type has its own parameter set, I have not knowledge enough to refactor this bit of code making easier to read and understand.

toni-moreno commented:
Hi @danielnelson, @russorat  I've been outside for some weeks sorry for the delay in the response, I will review your change request ASAP.

toni-moreno commented:
Hi @danielnelson , @russorat . 

I've finally completed the requested changes, also Rebased and added TLS refactor.
I'll be happy If you could give me more feedback ( if needed) . Thank you very much!

toni-moreno commented:
Hi @danielnelson @russorat .
I've reviewed the appveyor test and the error seems to be related to connetivity issues and not the code itself.
Could I do something to restart this test?

![image](https://user-images.githubusercontent.com/5883405/40163535-5815e0de-59b7-11e8-926f-b5ae531ba9d9.png)

toni-moreno commented:
@danielnelson  any news?

toni-moreno commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1169 @@
+		return fmt.Errorf("Error decoding JSON response: %s : %s", out, err)
+	}
+
+	errorChannel := make(chan error, len(deployments.Result))
```
ok

toni-moreno commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1169 @@
+
+// Gather Gathers data for all servers.
+func (h *JBoss) Gather(acc telegraf.Accumulator) error {
+	var wg sync.WaitGroup
```
ok

toni-moreno commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1169 @@
+			defer wg.Done()
+			//default as standalone server
+			hosts := HostResponse{Outcome: "", Result: []string{"standalone"}}
+			log.Printf("I! JBoss Plugin Working as Domain: %t\n", h.ExecAsDomain)
```
ok

toni-moreno commented:
Hi @danielnelson , sorry for the last review request  I didn't see some of your inicial comments, I think now all is ready to merge

danielnelson commented:
I'll take a look as soon as we wrap up 1.7, but I will need to push this until 1.8.

toni-moreno commented:
Hi @danielnelson I've committed a new feature which adds ability to autodiscover the execution mode  (standalone/domain) , and also improved  the error logging.

glinton commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1289 @@
+
+	fields := make(map[string]interface{})
+	for key, value := range server.Result {
+		log.Printf("D! LISTERNER %s : %s \n", key, value)
```
typo `listener`

glinton commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1289 @@
+}
+
+// Gathers database data from a particular host
+// Parameters:
```
outdated parameters

glinton commented on plugins/inputs/jboss/jboss.go:
```diff
@@ -0,0 +1,1289 @@
+
+	resp, err := h.client.MakeRequest(req)
+	if err != nil {
+		//log.Printf("D! HTTP REQ:%#+v", req)
```
Remove this and other commented code from this function

glinton commented:
@toni-moreno when do you think you'll have time to get to that last bit of feedback?

toni-moreno commented:
hi @glinton  I  will review ASAP

pjonesIDBS commented:
Is this still being worked on? It would still be good to have a JBoss plugin available.

sjwang90 commented:
@toni-moreno Checking if this is something you'd still be interested addressing the comments. Otherwise we can see if anyone else in the community wants to take this over.

toni-moreno commented:
Hello @sjwang90 , I'm sorry i would like to help but  I have no enough time to spend coding in this issue right now.  

Let me know if anyone who continue with the plugin needs for information about what I coded some time ago, I will be pleased to help by sharing my knowledgement

srebhan commented:
So is there any volunteer to take over this PR?

stefa975 commented:
Sure, I got some time over.

srebhan commented:
@stefa975 do you want to get used to the code first or do you prefer me shooting the review at you? ;-)

stefa975 commented:
@srebhan I have an updated version in my fork that works on the latest telegraf release. So I'll fix the reviews and push a new version.

srebhan commented:
Cool, so please reference this PR in your new one so we can close it once yours is merged.

sjwang90 commented:
Let move forward with #8926
